packages/common/src/AncillaryDataParser.ts

```diff
@@ -0,0 +1,256 @@
+        break;
+      }
+    }
+    key = key.concat(keyValue[index].character);
+    index++;
+  }
+
+  // No column (:) delimiter found, but reached the end of keyValue pair:
+  if (index === keyValue.length) throw "Cannot parse key value pair: no column delimiter found!";
+
+  // Skip unescaped whitespace
+  index +=
+    keyValue.slice(index).findIndex(skipWhitespace) === -1
+      ? keyValue.slice(index).length
+      : keyValue.slice(index).findIndex(skipWhitespace);
+
+  while (index < keyValue.length) {
+    const skip =
+      keyValue.slice(index).findIndex(skipWhitespace) === -1 ? 0 : keyValue.slice(index).findIndex(skipWhitespace);
+
+    // There should be only one unescaped column (:) delimiter in the keyValue pair:
+    if (keyValue[index + skip].character === ":" && !keyValue[index + skip].escape)
+      throw "Cannot parse key value pair: multiple column delimiters found!";
+
+    value = value.concat(keyValue[index].character);
+    index++;
+  }
+  if (!key || !value) throw "Cannot parse key value pair!";
+
+  // First try parsing value as JSON object:
+  try {
+    return [key, JSON.parse(value)];
+  } catch (err) {
+    // Then parse as Number or return string value:
+    if (value === Number(value).toString()) {
+      return [key, Number(value)];
+    } else {
+      return [key, value];
+    }
+  }
+}
+
+// Checks if reached end/start without whitespace.
+function isNextEnd(stringObject: CharObject[], start: number, forward = true): boolean {
+  if (forward) {
+    return stringObject.slice(start + 1).findIndex(skipWhitespace) === -1;
+  } else {
+    return stringObject.slice(0, start).reverse().findIndex(skipWhitespace) === -1;
+  }
+}
+
+// Checks if next non-whitespace character forward/backward matches the provided input character.
+export function isNextChar(stringObject: CharObject[], start: number, character: string, forward = true): boolean {
+  if (forward) {
+    const nextCharIndex = stringObject.slice(start + 1).findIndex(skipWhitespace);
+    if (nextCharIndex === -1) {
+      return false;
+    } else {
+      return (
+        stringObject[start + 1 + nextCharIndex].character === character &&
+        !stringObject[start + 1 + nextCharIndex].escape
+      );
+    }
+  } else {
+    const nextCharIndex = stringObject.slice(0, start).reverse().findIndex(skipWhitespace);
+    if (nextCharIndex === -1) {
+      return false;
+    } else {
+      return (
+        stringObject[start - 1 - nextCharIndex].character === character &&
+        !stringObject[start - 1 - nextCharIndex].escape
+      );
+    }
+  }
+}
+
+/**
+ * Finds closing quotes for keys/values and marks escaped.
+ * For values: closing quotes should be either before comma (,) or at the end.
+ * For keys: closing quotes should be before column (:).
+ */
+function escapeQuotes(stringObject: CharObject[], openIndex: number, escapeValues = true) {
+  const nextCharFn = escapeValues
+    ? function (stringObject: CharObject[], closeIndex: number) {
```

Comment:
Just learning myself on TS - why do we not set return type of boolean for this function? The same question applies to escaping keys (escapeValues = false) case below and `skipWhitespace` and `removeDoubleQuotes` that returns boolean.

Reply:
Good catch! Making the return values more explicit.